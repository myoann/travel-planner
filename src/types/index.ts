type TLanguage = {
  code: string;
  name: string;
  native: string;
  rtl: boolean;
};

type TSubdivision = {
  code: string;
  emoji: string;
  name: string;
};

type TState = {
  code: string;
};

type TContinent = {
  code: string;
  countries: {
    code: string;
  }[];
  name: string;
};

export type TCountry = {
  awsRegion: string;
  capital: string;
  code: string;
  continent: TContinent;
  currencies: string[];
  currency: string;
  emoji: string;
  emojiU: string;
  languages: TLanguage[];
  name: string;
  native: string;
  phone: string;
  phones: string[];
  states: TState[];
  subdivisions: TSubdivision[];
};

export type TOption = {
  key: string;
  name: string;
  value: string;
};
