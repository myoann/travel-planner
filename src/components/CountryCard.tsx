import { ChangeEvent } from "react";
import Image from "next/image";

import { TCountry } from "@/types";

type TProps = {
  country: TCountry;
  isChecked: boolean;
  isDisabled: boolean;
  onCompare: (event: ChangeEvent<HTMLInputElement>) => void;
};

export default function CountryCard({
  country: { code, name },
  isChecked,
  isDisabled,
  onCompare,
}: TProps) {
  return (
    <div className="flex flex-col bg-white rounded pb-4">
      <a href={`/country/${code}`} className="w-[200px] flex flex-col gap-2">
        <div className="relative w-full h-[150px]">
          <Image
            src={`/flags/${code.toLowerCase()}.svg`}
            alt={name}
            width={200}
            height={150}
            className="relative z-10 rounded-tl rounded-tr"
          />
          <div className="absolute inset-0 bg-gradient-to-t from-black/25 to-transparent z-20"></div>
        </div>

        <p className="text-wrap px-4 font-semibold line-clamp-2 h-12">{name}</p>
      </a>

      <label className="flex items-center mx-4 mt-2">
        <input
          type="checkbox"
          checked={isChecked}
          disabled={isDisabled}
          onChange={onCompare}
          className="h-6 w-6 text-[#1d1d1f] rounded border-gray-300"
        />

        <span className="ml-2 text-sm font-medium text-[#1d1d1f]">Compare</span>
      </label>
    </div>
  );
}
