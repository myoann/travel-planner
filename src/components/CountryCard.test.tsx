import { render, screen } from "@testing-library/react";

import { TCountry } from "@/types";

import CountryCard from "./CountryCard";

const defaultProps = {
  country: {
    code: "FR",
    name: "France",
  } as TCountry,
  isChecked: false,
  isDisabled: false,
  onCompare: jest.fn(),
};

describe("CountryCard", () => {
  it("matches the snapshot", () => {
    const { container } = render(<CountryCard {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it("has a a link which redirects to the country page", () => {
    render(<CountryCard {...defaultProps} />);

    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", "/country/FR");
  });

  it("displays an image and the name of the country", () => {
    render(<CountryCard {...defaultProps} />);

    const image = screen.getByRole("img", { name: "France" });
    const name = screen.getByText("France");

    expect(image).toBeInTheDocument();
    expect(name).toBeInTheDocument();
  });

  it("displays a checkbox to compare the country card", () => {
    render(<CountryCard {...defaultProps} />);

    const input = screen.getByRole("checkbox", { name: "Compare" });

    expect(input).toBeInTheDocument();
  });
});
