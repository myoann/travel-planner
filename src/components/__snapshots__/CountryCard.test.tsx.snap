// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryCard matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-col bg-white rounded pb-4"
  >
    <a
      class="md:w-[200px] sm:w-[133px] flex flex-col gap-2"
      href="/country/FR"
    >
      <div
        class="relative w-full md:h-[150px] sm:h-[100px]"
      >
        <img
          alt="France"
          class="relative z-10 rounded-tl rounded-tr"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/flags/fr.svg"
          style="color: transparent;"
          width="200"
        />
        <div
          class="absolute inset-0 bg-gradient-to-t from-black/25 to-transparent z-20"
        />
      </div>
      <p
        class="text-wrap px-4 font-semibold line-clamp-2 h-12"
      >
        France
      </p>
    </a>
    <label
      class="flex items-center mx-4 mt-2"
    >
      <input
        class="h-6 w-6 text-[#1d1d1f] rounded border-gray-300"
        type="checkbox"
      />
      <span
        class="ml-2 text-sm font-medium text-[#1d1d1f]"
      >
        Compare
      </span>
    </label>
  </div>
</div>
`;
